app/services/metrics/base_metric_processor.rb

```diff
@@ -0,0 +1,63 @@
+module Metrics
+  ##
+  # Base class for classes that do the processing of a Metric from a
+  # collection of events.
+  class BaseMetricProcessor < BaseService
+    attr_reader :events, :time_interval
+
+    def initialize(events:, time_interval:)
+      @events = events
+      @time_interval = time_interval
+    end
+
+    ##
+    # Processes the given events to generate the metrics.
+    # The given @events are expected to include all the events during the full
+    # time interval ( [@starting_at, @starting_at + @time_span) ).
+    def call
+      process_events
+    end
+
+    private
+
+    ##
+    # Processes the given events to generate the review_turnaround metrics.
+    def process_events
+      iterate_events
+      update_metrics
```

Comment:
it'd be good to define these 2 methods and raise an exception `Not Implemented` in the base metric processor

Reply:
The meaning of NotImplemented in Ruby is different from Python and other languages: [https://ruby-doc.org/core-2.5.0/NotImplementedError.html](https://ruby-doc.org/core-2.5.0/NotImplementedError.html).

Also [this](https://ruby-doc.org/core-2.5.0/NotImplementedError.html) style guide suggest to prefer duck tape.